File: intervention/prototypes/teacher_dashboard/src/shared/data/teacherStorage.ts
export interface TeacherProfile {
  name: string;
  email: string;
  school: string;
  avatar: string;
}

export interface Student {
  id: string;
  name: string;
  avatar: string;
  classId: string;
  progressPercentage: number;
  challengesCompleted: number;
  lessonsCompleted: number;
  lastActivity: string;
  achievements: Achievement[];
  activityHistory: ActivityEntry[];
  teacherNotes: string;
}

export interface Achievement {
  id: string;
  name: string;
  icon: string;
  earnedDate: string;
}

export interface ActivityEntry {
  id: string;
  type: 'lesson' | 'challenge' | 'achievement';
  title: string;
  date: string;
  success: boolean;
}

export interface Class {
  id: string;
  name: string;
  studentCount: number;
  activeAssignments: number;
  engagementRate: number;
  students: string[]; // student IDs
}

export interface Assignment {
  id: string;
  lessonId: string;
  classId: string;
  assignedDate: string;
  dueDate: string;
  completionRate: number;
}

const STORAGE_KEYS = {
  TEACHER_PROFILE: 'teacher_profile',
  CLASSES: 'teacher_classes',
  STUDENTS: 'teacher_students',
  ASSIGNMENTS: 'teacher_assignments',
  LAST_LESSON: 'teacher_last_lesson'
};

// Teacher Profile
export function saveTeacherProfile(profile: TeacherProfile): void {
  localStorage.setItem(STORAGE_KEYS.TEACHER_PROFILE, JSON.stringify(profile));
}

export function getTeacherProfile(): TeacherProfile | null {
  const data = localStorage.getItem(STORAGE_KEYS.TEACHER_PROFILE);
  return data ? JSON.parse(data) : null;
}

// Classes
export function getClasses(): Class[] {
  const data = localStorage.getItem(STORAGE_KEYS.CLASSES);
  if (data) {
    return JSON.parse(data);
  }

  // Default mock classes
  const defaultClasses: Class[] = [
    {
      id: 'class-1',
      name: 'Class 4A',
      studentCount: 24,
      activeAssignments: 3,
      engagementRate: 87,
      students: ['student-1', 'student-2', 'student-3', 'student-4', 'student-5']
    },
    {
      id: 'class-2',
      name: 'Class 4B',
      studentCount: 22,
      activeAssignments: 2,
      engagementRate: 92,
      students: ['student-6', 'student-7', 'student-8', 'student-9']
    },
    {
      id: 'class-3',
      name: 'Class 5A',
      studentCount: 26,
      activeAssignments: 4,
      engagementRate: 78,
      students: ['student-10', 'student-11', 'student-12']
    }
  ];

  localStorage.setItem(STORAGE_KEYS.CLASSES, JSON.stringify(defaultClasses));
  return defaultClasses;
}

export function addClass(newClass: Class): void {
  const classes = getClasses();
  classes.push(newClass);
  localStorage.setItem(STORAGE_KEYS.CLASSES, JSON.stringify(classes));
}

// Students
export function getStudents(): Student[] {
  const data = localStorage.getItem(STORAGE_KEYS.STUDENTS);
  if (data) {
    return JSON.parse(data);
  }

  // Default mock students
  const defaultStudents: Student[] = [
    {
      id: 'student-1',
      name: 'Emma de Vries',
      avatar: '👧',
      classId: 'class-1',
      progressPercentage: 78,
      challengesCompleted: 15,
      lessonsCompleted: 4,
      lastActivity: new Date(Date.now() - 2 * 60 * 60 * 1000).toISOString(),
      achievements: [
        { id: 'ach-1', name: 'Pattern Master', icon: '🎯', earnedDate: new Date(Date.now() - 5 * 24 * 60 * 60 * 1000).toISOString() },
        { id: 'ach-2', name: 'Loop Hero', icon: '🔄', earnedDate: new Date(Date.now() - 3 * 24 * 60 * 60 * 1000).toISOString() }
      ],
      activityHistory: [
        { id: 'act-1', type: 'lesson', title: 'Understanding Loops', date: new Date(Date.now() - 2 * 60 * 60 * 1000).toISOString(), success: true },
        { id: 'act-2', type: 'challenge', title: 'Loop Challenge', date: new Date(Date.now() - 1 * 24 * 60 * 60 * 1000).toISOString(), success: true }
      ],
      teacherNotes: 'Very engaged student, excels at pattern recognition'
    },
    {
      id: 'student-2',
      name: 'Lucas van Berg',
      avatar: '👦',
      classId: 'class-1',
      progressPercentage: 92,
      challengesCompleted: 22,
      lessonsCompleted: 6,
      lastActivity: new Date(Date.now() - 5 * 60 * 60 * 1000).toISOString(),
      achievements: [
        { id: 'ach-1', name: 'Pattern Master', icon: '🎯', earnedDate: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString() },
        { id: 'ach-2', name: 'Loop Hero', icon: '🔄', earnedDate: new Date(Date.now() - 6 * 24 * 60 * 60 * 1000).toISOString() },
        { id: 'ach-3', name: 'Variable Wizard', icon: '✨', earnedDate: new Date(Date.now() - 2 * 24 * 60 * 60 * 1000).toISOString() }
      ],
      activityHistory: [
        { id: 'act-3', type: 'lesson', title: 'Variables and Storage', date: new Date(Date.now() - 5 * 60 * 60 * 1000).toISOString(), success: true },
        { id: 'act-4', type: 'achievement', title: 'Variable Wizard', date: new Date(Date.now() - 2 * 24 * 60 * 60 * 1000).toISOString(), success: true }
      ],
      teacherNotes: 'Top performer, helps other students'
    },
    {
      id: 'student-3',
      name: 'Sophie Bakker',
      avatar: '👧',
      classId: 'class-1',
      progressPercentage: 64,
      challengesCompleted: 12,
      lessonsCompleted: 3,
      lastActivity: new Date(Date.now() - 1 * 24 * 60 * 60 * 1000).toISOString(),
      achievements: [
        { id: 'ach-1', name: 'Pattern Master', icon: '🎯', earnedDate: new Date(Date.now() - 4 * 24 * 60 * 60 * 1000).toISOString() }
      ],
      activityHistory: [
        { id: 'act-5', type: 'lesson', title: 'Introduction to Patterns', date: new Date(Date.now() - 1 * 24 * 60 * 60 * 1000).toISOString(), success: true }
      ],
      teacherNotes: 'Needs extra support with conditional logic'
    },
    {
      id: 'student-4',
      name: 'Daan Jansen',
      avatar: '👦',
      classId: 'class-1',
      progressPercentage: 85,
      challengesCompleted: 18,
      lessonsCompleted: 5,
      lastActivity: new Date(Date.now() - 3 * 60 * 60 * 1000).toISOString(),
      achievements: [
        { id: 'ach-1', name: 'Pattern Master', icon: '🎯', earnedDate: new Date(Date.now() - 8 * 24 * 60 * 60 * 1000).toISOString() },
        { id: 'ach-2', name: 'Loop Hero', icon: '🔄', earnedDate: new Date(Date.now() - 5 * 24 * 60 * 60 * 1000).toISOString() }
      ],
      activityHistory: [
        { id: 'act-6', type: 'lesson', title: 'Conditional Logic', date: new Date(Date.now() - 3 * 60 * 60 * 1000).toISOString(), success: true }
      ],
      teacherNotes: 'Creative problem solver'
    },
    {
      id: 'student-5',
      name: 'Lotte Peters',
      avatar: '👧',
      classId: 'class-1',
      progressPercentage: 71,
      challengesCompleted: 14,
      lessonsCompleted: 4,
      lastActivity: new Date(Date.now() - 8 * 60 * 60 * 1000).toISOString(),
      achievements: [
        { id: 'ach-1', name: 'Pattern Master', icon: '🎯', earnedDate: new Date(Date.now() - 6 * 24 * 60 * 60 * 1000).toISOString() },
        { id: 'ach-2', name: 'Loop Hero', icon: '🔄', earnedDate: new Date(Date.now() - 4 * 24 * 60 * 60 * 1000).toISOString() }
      ],
      activityHistory: [
        { id: 'act-7', type: 'challenge', title: 'Loop Builder', date: new Date(Date.now() - 8 * 60 * 60 * 1000).toISOString(), success: true }
      ],
      teacherNotes: 'Consistent progress, attentive in class'
    }
  ];

  localStorage.setItem(STORAGE_KEYS.STUDENTS, JSON.stringify(defaultStudents));
  return defaultStudents;
}

export function getStudentsByClass(classId: string): Student[] {
  const students = getStudents();
  return students.filter(s => s.classId === classId);
}

export function getStudentById(studentId: string): Student | null {
  const students = getStudents();
  return students.find(s => s.id === studentId) || null;
}

export function updateStudentNotes(studentId: string, notes: string): void {
  const students = getStudents();
  const student = students.find(s => s.id === studentId);
  if (student) {
    student.teacherNotes = notes;
    localStorage.setItem(STORAGE_KEYS.STUDENTS, JSON.stringify(students));
  }
}

// Assignments
export function getAssignments(): Assignment[] {
  const data = localStorage.getItem(STORAGE_KEYS.ASSIGNMENTS);
  if (data) {
    return JSON.parse(data);
  }
  return [];
}

export function createAssignment(assignment: Assignment): void {
  const assignments = getAssignments();
  assignments.push(assignment);
  localStorage.setItem(STORAGE_KEYS.ASSIGNMENTS, JSON.stringify(assignments));
}

// Last viewed lesson
export function saveLastLesson(lessonId: string): void {
  localStorage.setItem(STORAGE_KEYS.LAST_LESSON, lessonId);
}

export function getLastLesson(): string | null {
  return localStorage.getItem(STORAGE_KEYS.LAST_LESSON);
}

File: intervention/prototypes/teacher_dashboard_python/data/teacher_storage.py
import json
import os
from pathlib import Path
from datetime import datetime

# Data directory
DATA_DIR = Path(__file__).parent / 'store'
DATA_DIR.mkdir(exist_ok=True)

STORAGE_KEYS = {
    'TEACHER_PROFILE': DATA_DIR / 'teacher_profile.json',
    'CLASSES': DATA_DIR / 'classes.json',
    'STUDENTS': DATA_DIR / 'students.json',
    'ASSIGNMENTS': DATA_DIR / 'assignments.json',
    'LAST_LESSON': DATA_DIR / 'last_lesson.txt'
}

# Default data
DEFAULT_TEACHER_PROFILE = {
    'name': 'Sarah Johnson',
    'email': 'sarah.johnson@school.edu',
    'school': 'Kennedy Elementary School',
    'avatar': 'SJ'
}

DEFAULT_CLASSES = [
    {
        'id': 'class-1',
        'name': 'Class 4A',
        'studentCount': 24,
        'activeAssignments': 3,
        'engagementRate': 87,
        'students': ['student-1', 'student-2', 'student-3', 'student-4', 'student-5']
    },
    {
        'id': 'class-2',
        'name': 'Class 4B',
        'studentCount': 22,
        'activeAssignments': 2,
        'engagementRate': 92,
        'students': ['student-6', 'student-7', 'student-8', 'student-9']
    },
    {
        'id': 'class-3',
        'name': 'Class 5A',
        'studentCount': 20,
        'activeAssignments': 4,
        'engagementRate': 78,
        'students': ['student-10', 'student-11']
    }
]

DEFAULT_STUDENTS = [
    {
        'id': 'student-1',
        'name': 'Emma Wilson',
        'avatar': 'EW',
        'classId': 'class-1',
        'progressPercentage': 75,
        'challengesCompleted': 12,
        'lessonsCompleted': 5,
        'lastActivity': '2 hours ago',
        'achievements': [
            {'id': 'ach-1', 'name': 'First Steps', 'icon': 'star', 'earnedDate': '2024-01-15'},
            {'id': 'ach-2', 'name': 'Quick Learner', 'icon': 'zap', 'earnedDate': '2024-02-20'}
        ],
        'activityHistory': [
            {'id': 'act-1', 'type': 'lesson', 'title': 'Loops & Sequences', 'date': '2024-03-15', 'success': True},
            {'id': 'act-2', 'type': 'challenge', 'title': 'Pattern Challenge', 'date': '2024-03-10', 'success': True}
        ],
        'teacherNotes': 'Emma shows great promise. Very attentive during class.'
    },
    {
        'id': 'student-2',
        'name': 'Liam Chen',
        'avatar': 'LC',
        'classId': 'class-1',
        'progressPercentage': 82,
        'challengesCompleted': 15,
        'lessonsCompleted': 7,
        'lastActivity': '30 minutes ago',
        'achievements': [
            {'id': 'ach-1', 'name': 'First Steps', 'icon': 'star', 'earnedDate': '2024-01-10'}
        ],
        'activityHistory': [
            {'id': 'act-1', 'type': 'lesson', 'title': 'Variables & Data Types', 'date': '2024-03-16', 'success': True}
        ],
        'teacherNotes': 'Excellent progress. Helping other students.'
    },
    {
        'id': 'student-3',
        'name': 'Marcus Brown',
        'avatar': 'MB',
        'classId': 'class-1',
        'progressPercentage': 45,
        'challengesCompleted': 6,
        'lessonsCompleted': 2,
        'lastActivity': '1 day ago',
        'achievements': [],
        'activityHistory': [],
        'teacherNotes': 'Needs additional support. Meeting scheduled.'
    },
    {
        'id': 'student-4',
        'name': 'Sofia Garcia',
        'avatar': 'SG',
        'classId': 'class-1',
        'progressPercentage': 88,
        'challengesCompleted': 18,
        'lessonsCompleted': 8,
        'lastActivity': '1 hour ago',
        'achievements': [
            {'id': 'ach-1', 'name': 'First Steps', 'icon': 'star', 'earnedDate': '2024-01-05'},
            {'id': 'ach-2', 'name': 'Quick Learner', 'icon': 'zap', 'earnedDate': '2024-02-15'},
            {'id': 'ach-3', 'name': 'Problem Solver', 'icon': 'lightbulb', 'earnedDate': '2024-03-01'}
        ],
        'activityHistory': [],
        'teacherNotes': 'Outstanding student. Role model for others.'
    },
    {
        'id': 'student-5',
        'name': 'James Park',
        'avatar': 'JP',
        'classId': 'class-1',
        'progressPercentage': 68,
        'challengesCompleted': 10,
        'lessonsCompleted': 4,
        'lastActivity': '45 minutes ago',
        'achievements': [
            {'id': 'ach-1', 'name': 'First Steps', 'icon': 'star', 'earnedDate': '2024-01-20'}
        ],
        'activityHistory': [],
        'teacherNotes': 'Good progress. Enjoys visual learning.'
    },
    {
        'id': 'student-6',
        'name': 'Olivia Martinez',
        'avatar': 'OM',
        'classId': 'class-2',
        'progressPercentage': 92,
        'challengesCompleted': 20,
        'lessonsCompleted': 9,
        'lastActivity': '20 minutes ago',
        'achievements': [
            {'id': 'ach-1', 'name': 'First Steps', 'icon': 'star', 'earnedDate': '2024-01-02'},
            {'id': 'ach-2', 'name': 'Quick Learner', 'icon': 'zap', 'earnedDate': '2024-02-08'}
        ],
        'activityHistory': [],
        'teacherNotes': 'Class leader in coding challenges.'
    },
    {
        'id': 'student-7',
        'name': 'Noah Johnson',
        'avatar': 'NJ',
        'classId': 'class-2',
        'progressPercentage': 61,
        'challengesCompleted': 9,
        'lessonsCompleted': 3,
        'lastActivity': '3 hours ago',
        'achievements': [],
        'activityHistory': [],
        'teacherNotes': 'Progressing steadily. Prefers group work.'
    },
    {
        'id': 'student-8',
        'name': 'Ava Taylor',
        'avatar': 'AT',
        'classId': 'class-2',
        'progressPercentage': 79,
        'challengesCompleted': 14,
        'lessonsCompleted': 6,
        'lastActivity': '2 hours ago',
        'achievements': [
            {'id': 'ach-1', 'name': 'First Steps', 'icon': 'star', 'earnedDate': '2024-01-12'}
        ],
        'activityHistory': [],
        'teacherNotes': 'Consistent high quality work.'
    },
    {
        'id': 'student-9',
        'name': 'Noah Lee',
        'avatar': 'NL',
        'classId': 'class-2',
        'progressPercentage': 55,
        'challengesCompleted': 7,
        'lessonsCompleted': 2,
        'lastActivity': '5 hours ago',
        'achievements': [],
        'activityHistory': [],
        'teacherNotes': 'Needs encouragement. New to coding.'
    },
    {
        'id': 'student-10',
        'name': 'Grace Kim',
        'avatar': 'GK',
        'classId': 'class-3',
        'progressPercentage': 85,
        'challengesCompleted': 16,
        'lessonsCompleted': 7,
        'lastActivity': '1 hour ago',
        'achievements': [
            {'id': 'ach-1', 'name': 'First Steps', 'icon': 'star', 'earnedDate': '2024-01-08'}
        ],
        'activityHistory': [],
        'teacherNotes': 'Creative problem solver.'
    },
    {
        'id': 'student-11',
        'name': 'Henry Davis',
        'avatar': 'HD',
        'classId': 'class-3',
        'progressPercentage': 71,
        'challengesCompleted': 11,
        'lessonsCompleted': 5,
        'lastActivity': '4 hours ago',
        'achievements': [],
        'activityHistory': [],
        'teacherNotes': 'Making good progress.'
    }
]

DEFAULT_ASSIGNMENTS = [
    {
        'id': 'assign-1',
        'lessonId': 'lesson-1',
        'classId': 'class-1',
        'assignedDate': '2024-03-10',
        'dueDate': '2024-03-17',
        'completionRate': 85
    },
    {
        'id': 'assign-2',
        'lessonId': 'lesson-2',
        'classId': 'class-1',
        'assignedDate': '2024-03-12',
        'dueDate': '2024-03-19',
        'completionRate': 70
    }
]

# Teacher Profile
def get_teacher_profile():
    """Get teacher profile from storage"""
    if STORAGE_KEYS['TEACHER_PROFILE'].exists():
        with open(STORAGE_KEYS['TEACHER_PROFILE'], 'r') as f:
            return json.load(f)
    return DEFAULT_TEACHER_PROFILE

def save_teacher_profile(profile):
    """Save teacher profile to storage"""
    with open(STORAGE_KEYS['TEACHER_PROFILE'], 'w') as f:
        json.dump(profile, f, indent=2)

# Classes
def get_classes():
    """Get classes from storage"""
    if STORAGE_KEYS['CLASSES'].exists():
        with open(STORAGE_KEYS['CLASSES'], 'r') as f:
            return json.load(f)
    return DEFAULT_CLASSES

def save_classes(classes):
    """Save classes to storage"""
    with open(STORAGE_KEYS['CLASSES'], 'w') as f:
        json.dump(classes, f, indent=2)

# Students
def get_students():
    """Get students from storage"""
    if STORAGE_KEYS['STUDENTS'].exists():
        with open(STORAGE_KEYS['STUDENTS'], 'r') as f:
            return json.load(f)
    return DEFAULT_STUDENTS

def save_students(students):
    """Save students to storage"""
    with open(STORAGE_KEYS['STUDENTS'], 'w') as f:
        json.dump(students, f, indent=2)

# Assignments
def get_assignments():
    """Get assignments from storage"""
    if STORAGE_KEYS['ASSIGNMENTS'].exists():
        with open(STORAGE_KEYS['ASSIGNMENTS'], 'r') as f:
            return json.load(f)
    return DEFAULT_ASSIGNMENTS

def save_assignments(assignments):
    """Save assignments to storage"""
    with open(STORAGE_KEYS['ASSIGNMENTS'], 'w') as f:
        json.dump(assignments, f, indent=2)

# Last Lesson
def get_last_lesson():
    """Get last viewed lesson ID"""
    if STORAGE_KEYS['LAST_LESSON'].exists():
        with open(STORAGE_KEYS['LAST_LESSON'], 'r') as f:
            return f.read().strip()
    return None

def save_last_lesson(lesson_id):
    """Save last viewed lesson ID"""
    with open(STORAGE_KEYS['LAST_LESSON'], 'w') as f:
        f.write(lesson_id)
